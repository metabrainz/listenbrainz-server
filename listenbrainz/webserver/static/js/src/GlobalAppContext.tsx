import { createContext } from "react";
import APIService from "./APIService";

export type GlobalAppContextT = {
  APIService: APIService;
  currentUser?: ListenBrainzUser;
  spotifyAuth?: SpotifyUser;
  youtubeAuth?: {};
};

const GlobalAppContext = createContext<GlobalAppContextT>({
  APIService: new APIService(`${window.location.origin}/1`),
  currentUser: {} as ListenBrainzUser,
  spotifyAuth: {} as SpotifyUser,
  youtubeAuth: {},
});

export default GlobalAppContext;
