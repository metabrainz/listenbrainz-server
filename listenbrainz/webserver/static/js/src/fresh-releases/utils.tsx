import { useEffect, useState } from "react";

export function formatReleaseDate(releaseDate: string) {
  return new Intl.DateTimeFormat("en-US", {
    month: "short",
    day: "numeric",
  })
    .formatToParts(new Date(Date.parse(releaseDate)))
    .reverse()
    .map((date_parts) => date_parts.value)
    .join("");
}

// Originally from https://usehooks-ts.com/react-hook/use-media-query
export function useMediaQuery(queryStr: string) {
  const getMatches = (query: string): boolean => {
    if (typeof window !== "undefined") {
      return window.matchMedia(query).matches;
    }
    return false;
  };

  const [matches, setMatches] = useState<boolean>(getMatches(queryStr));

  function handleChange() {
    setMatches(getMatches(queryStr));
  }

  useEffect(() => {
    const matchMedia = window.matchMedia(queryStr);
    handleChange();
    matchMedia.addEventListener("change", handleChange);
    return () => {
      matchMedia.removeEventListener("change", handleChange);
    };
  }, [queryStr]);

  return matches;
}
