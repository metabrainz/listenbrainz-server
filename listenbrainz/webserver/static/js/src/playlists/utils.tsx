/* eslint-disable camelcase */

import { padStart } from "lodash";

export const MUSICBRAINZ_JSPF_PLAYLIST_EXTENSION =
  "https://musicbrainz.org/doc/jspf#playlist";

export const MUSICBRAINZ_JSPF_TRACK_EXTENSION =
  "https://musicbrainz.org/doc/jspf#track";

export const PLAYLIST_URI_PREFIX = "https://listenbrainz.org/playlist/";
export const PLAYLIST_TRACK_URI_PREFIX = "https://musicbrainz.org/recording/";

export function getPlaylistExtension(
  playlist?: JSPFPlaylist
): JSPFPlaylistExtension | null {
  if (!playlist) {
    return null;
  }
  return playlist.extension?.[MUSICBRAINZ_JSPF_PLAYLIST_EXTENSION] ?? null;
}

export function getTrackExtension(
  track?: JSPFTrack
): JSPFTrackExtension | null {
  if (!track) {
    return null;
  }
  return track.extension?.[MUSICBRAINZ_JSPF_TRACK_EXTENSION] ?? null;
}

export function getPlaylistId(playlist?: JSPFPlaylist): string {
  return playlist?.identifier?.substr(PLAYLIST_URI_PREFIX.length) ?? "";
}

export function getRecordingMBIDFromJSPFTrack(track: JSPFTrack): string {
  return track.identifier?.substr(PLAYLIST_TRACK_URI_PREFIX.length) ?? "";
}

// Credit goes to Dmitry Sheiko https://stackoverflow.com/a/53006402/4904467
export function millisecondsToStr(milliseconds: number) {
  let temp = milliseconds / 1000;
  const days = Math.floor((temp %= 31536000) / 86400);
  const hours = Math.floor((temp %= 86400) / 3600);
  const minutes = Math.floor((temp %= 3600) / 60);
  const seconds = temp % 60;

  if (seconds > 1 && !minutes && !hours) {
    return `${seconds.toFixed(0)}s`;
  }
  if (hours || minutes) {
    return `${
      (days ? `${days}d ` : "") +
      (hours ? `${hours}:` : "") +
      (minutes ? `${minutes}:` : "") +
      padStart(seconds.toFixed(0), 2, "0")
    }`;
  }

  return "< 1s";
}

export function JSPFTrackToListen(track: JSPFTrack): Listen {
  return {
    listened_at: 0,
    track_metadata: {
      artist_name: track.creator,
      track_name: track.title,
      release_name: track.album,
      additional_info: {
        duration_ms: track.duration,
        recording_mbid: track.id,
        origin_url: track.location?.[0],
      },
    },
  };
}

export function listenToJSPFTrack(listen: Listen): JSPFTrack {
  return {
    identifier:
      PLAYLIST_TRACK_URI_PREFIX +
      listen.track_metadata?.additional_info?.recording_mbid,
    id: listen.track_metadata?.additional_info?.recording_mbid || undefined,
    title: listen.track_metadata?.track_name,
    creator: listen.track_metadata?.artist_name,
    album: listen.track_metadata?.release_name || undefined,
    duration: listen.track_metadata?.additional_info?.duration_ms || undefined,
    location: listen.track_metadata?.additional_info?.origin_url
      ? [listen.track_metadata?.additional_info?.origin_url]
      : undefined,
  };
}
