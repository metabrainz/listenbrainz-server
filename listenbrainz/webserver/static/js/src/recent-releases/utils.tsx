const getCoverartFromReleaseMBID = async (
  releaseMBID: string
): Promise<string | undefined> => {
  try {
    const CAAResponse = await fetch(
      `https://coverartarchive.org/release/${releaseMBID}`
    );
    if (CAAResponse.ok) {
      const body: CoverArtArchiveResponse = await CAAResponse.json();
      if (!body.images?.[0]?.thumbnails) {
        return undefined;
      }
      const { thumbnails } = body.images[0];
      return (
        thumbnails[250] ??
        thumbnails.small ??
        // If neither of the above exists, return the first one we find
        // @ts-ignore
        thumbnails[Object.keys(thumbnails)?.[0]]
      );
    }
  } catch (error) {
    // eslint-disable-next-line no-console
    console.warn(
      `Couldn't fetch Cover Art Archive entry for ${releaseMBID}`,
      error
    );
  }
  return undefined;
};

const formattedReleaseDate = (releaseDate: string) => {
  return new Intl.DateTimeFormat("default", {
    month: "short",
    day: "numeric",
  })
    .formatToParts(new Date(Date.parse(releaseDate)))
    .reverse()
    .map((date_parts) => date_parts.value)
    .join("");
};

export { getCoverartFromReleaseMBID, formattedReleaseDate };
