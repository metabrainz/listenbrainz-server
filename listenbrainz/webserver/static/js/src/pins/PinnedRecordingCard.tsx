import * as React from "react";
import { isEqual, isNil } from "lodash";
import MediaQuery from "react-responsive";
import { faEllipsisV, faThumbtack } from "@fortawesome/free-solid-svg-icons";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import {
  preciseTimestamp,
  getListenablePin,
  pinnedRecordingToListen,
} from "../utils/utils";
import GlobalAppContext from "../utils/GlobalAppContext";
import Card from "../components/Card";
import ListenControl from "../listens/ListenControl";
import ListenCard from "../listens/ListenCard";

export const DEFAULT_COVER_ART_URL = "/static/img/default_cover_art.png";

export type PinnedRecordingCardProps = {
  userName: string;
  pinnedRecording: PinnedRecording;
  currentFeedback?: ListenFeedBack | null;
  // Only used when not passing a custom feedbackComponent
  updateFeedbackCallback?: (
    recordingMsid: string,
    score: ListenFeedBack | RecommendationFeedBack
  ) => void;
  className?: string;
  isCurrentUser: Boolean;
  removePinFromPinsList: (pin: PinnedRecording) => void;
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => void;
};

type PinnedRecordingCardState = {
  currentlyPinned?: Boolean;
  isDeleted: Boolean;
};

export default class PinnedRecordingCard extends React.Component<
  PinnedRecordingCardProps,
  PinnedRecordingCardState
> {
  static contextType = GlobalAppContext;
  declare context: React.ContextType<typeof GlobalAppContext>;

  constructor(props: PinnedRecordingCardProps) {
    super(props);
    this.state = {
      currentlyPinned: this.determineIfCurrentlyPinned(),
      isDeleted: false,
    };
  }

  determineIfCurrentlyPinned = (): Boolean => {
    const { pinnedRecording } = this.props;
    const pinnedUntilTime: Date = new Date(pinnedRecording.pinned_until * 1000);

    // invalid date
    if (Number.isNaN(pinnedUntilTime.getTime())) {
      return false;
    }
    const currentTime: Date = new Date();

    if (currentTime < pinnedUntilTime) {
      return true;
    }

    return false;
  };

  handleError = (error: string | Error, title?: string): void => {
    const { newAlert } = this.props;
    if (!error) {
      return;
    }
    newAlert(
      "danger",
      title || "Error",
      typeof error === "object" ? error.message : error
    );
  };

  unpinRecording = async () => {
    const { pinnedRecording, isCurrentUser, newAlert } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.unpinRecording(currentUser.auth_token);
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          newAlert(
            "success",
            `You unpinned a recording.`,
            `${pinnedRecording.track_metadata?.artist_name} - ${pinnedRecording.track_metadata?.track_name}`
          );
        }
      } catch (error) {
        this.handleError(error, "Error while unpinning recording");
      }
    }
  };

  deletePin = async (pin: PinnedRecording) => {
    const { isCurrentUser, removePinFromPinsList } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.deletePin(
          currentUser.auth_token,
          pin.row_id
        );
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          this.setState({ isDeleted: true });

          // wait for the animation to finish
          setTimeout(function removeListen() {
            removePinFromPinsList(pin);
          }, 1000);
        }
      } catch (error) {
        this.handleError(error, "Error while deleting pin");
      }
    }
  };

  render() {
    const {
      pinnedRecording,
      newAlert,
      currentFeedback,
      updateFeedbackCallback,
    } = this.props;
    const { currentlyPinned, isDeleted } = this.state;

    const thumbnail = currentlyPinned ? (
      <div className="pinned-recording-icon">
        <span className="icon">
          <FontAwesomeIcon icon={faThumbtack as IconProp} />
        </span>
        <span className="small">Pinned</span>
      </div>
    ) : undefined;

    const blurb = pinnedRecording.blurb_content ? (
      <div className="blurb-content" title={pinnedRecording.blurb_content}>
        &quot;{pinnedRecording.blurb_content}&quot;
      </div>
    ) : undefined;

    const additionalMenuItems = (
      <>
        {currentlyPinned && (
          <ListenControl
            title="Unpin"
            text="Unpin"
            action={() => this.unpinRecording()}
          />
        )}
        <ListenControl
          title="Delete Pin"
          text="Delete Pin"
          action={() => this.deletePin(pinnedRecording)}
        />
      </>
    );
    const cssClasses = ["pinned-recording-card"];
    if (currentlyPinned) {
      cssClasses.push("currently-pinned");
    }
    if (isDeleted) {
      cssClasses.push("deleted");
    }
    return (
      <ListenCard
        className={cssClasses.join(" ")}
        listen={pinnedRecordingToListen(pinnedRecording)}
        currentFeedback={currentFeedback}
        updateFeedbackCallback={updateFeedbackCallback}
        showTimestamp
        showUsername={false}
        newAlert={newAlert}
        additionalMenuItems={additionalMenuItems}
        additionalContent={blurb}
        thumbnail={thumbnail}
      />
    );
  }
}
