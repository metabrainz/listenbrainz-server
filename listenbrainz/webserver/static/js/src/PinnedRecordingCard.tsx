import * as React from "react";
import { get as _get } from "lodash";
import MediaQuery from "react-responsive";
import { faEllipsisV, faThumbtack } from "@fortawesome/free-solid-svg-icons";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { preciseTimestamp } from "./utils";
import GlobalAppContext from "./GlobalAppContext";
import Card from "./components/Card";
import ListenControl from "./listens/ListenControl";

export const DEFAULT_COVER_ART_URL = "/static/img/default_cover_art.png";

export type PinnedRecordingCardProps = {
  userName: string;
  pinnedRecording: PinnedRecording;
  isCurrentUser: Boolean;
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => void;
};

type PinnedRecordingCardState = {
  currentlyPinned?: Boolean;
  isDeleted: Boolean;
};

export default class PinnedRecordingCard extends React.Component<
  PinnedRecordingCardProps,
  PinnedRecordingCardState
> {
  static contextType = GlobalAppContext;
  declare context: React.ContextType<typeof GlobalAppContext>;

  constructor(props: PinnedRecordingCardProps) {
    super(props);
    const currentlyPinned = this.determineIfCurrentlyPinned();
    this.state = {
      currentlyPinned,
      isDeleted: false,
    };
  }

  determineIfCurrentlyPinned = (): Boolean => {
    const { pinnedRecording } = this.props;
    const pinnedUntilTime: Date = new Date(pinnedRecording.pinned_until * 1000);

    // invalid date
    if (Number.isNaN(pinnedUntilTime.getTime())) {
      return false;
    }
    const currentTime: Date = new Date();

    if (currentTime < pinnedUntilTime) {
      return true;
    }

    return false;
  };

  renderPinTitle = (): JSX.Element => {
    const { currentlyPinned } = this.state;
    const { isCurrentUser, pinnedRecording, userName } = this.props;

    const userText = isCurrentUser ? `${userName}'s` : "My";
    const { track_name } = pinnedRecording.track_metadata;

    return (
      <div>
        {currentlyPinned && (
          <p className="pin-title text-muted">
            <FontAwesomeIcon icon={faThumbtack as IconProp} /> {userText} Pinned
            Recording
          </p>
        )}
        <p title={track_name}>{track_name}</p>
      </div>
    );
  };

  renderPinDate = (): JSX.Element => {
    const { pinnedRecording } = this.props;
    return (
      <div
        className="pinned-recording-time text-center text-muted-less"
        title={new Date(pinnedRecording.created * 1000).toISOString()}
      >
        {preciseTimestamp(pinnedRecording.created * 1000)}
      </div>
    );
  };

  renderBlurbContent = (): JSX.Element | null => {
    const { pinnedRecording } = this.props;
    if (!pinnedRecording.blurb_content) {
      return null;
    }
    return (
      <div className="blurb-content" title={pinnedRecording.blurb_content}>
        &quot;{pinnedRecording.blurb_content}&quot;
      </div>
    );
  };

  handleError = (error: string | Error, title?: string): void => {
    const { newAlert } = this.props;
    if (!error) {
      return;
    }
    newAlert(
      "danger",
      title || "Error",
      typeof error === "object" ? error.message : error
    );
  };

  unpinRecording = async () => {
    const { pinnedRecording, isCurrentUser, newAlert } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.unpinRecording(currentUser.auth_token);
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          newAlert(
            "success",
            `You unpinned a recording.`,
            `${pinnedRecording.track_metadata.artist_name} - ${pinnedRecording.track_metadata.track_name}`
          );
        }
      } catch (error) {
        this.handleError(error, "Error while unpinning recording");
      }
    }
  };

  deletePin = async (pinID: number) => {
    const { isCurrentUser } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.deletePin(
          currentUser.auth_token,
          pinID
        );
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          this.setState({ isDeleted: true });
        }
      } catch (error) {
        this.handleError(error, "Error while deleting pin");
      }
    }
  };

  render() {
    const { pinnedRecording, isCurrentUser } = this.props;
    const { currentlyPinned, isDeleted } = this.state;
    const { artist_name } = pinnedRecording.track_metadata;

    return (
      <Card
        className={`pinned-recording-card row ${
          currentlyPinned ? "currently-pinned " : ""
        } ${isDeleted ? "deleted " : ""}`}
      >
        <div className={`${isCurrentUser ? " col-xs-9" : " col-xs-12"}`}>
          {/* Desktop browser layout */}
          <MediaQuery minWidth={768}>
            <div className="col-xs-8">
              <div className="pin-details">
                {this.renderPinTitle()}
                <small title={artist_name}>{artist_name}</small>
                {this.renderBlurbContent()}
              </div>
            </div>
            <div className="col-xs-4">{this.renderPinDate()}</div>
          </MediaQuery>

          {/* Mobile device layout */}
          <MediaQuery maxWidth={767}>
            <div className="col-xs-12">
              <div className="pin-details">
                {this.renderPinTitle()}
                <small title={artist_name}>
                  {this.renderPinDate()}
                  &nbsp; &#8212; &nbsp; {artist_name}
                </small>
                {this.renderBlurbContent()}
              </div>
            </div>
          </MediaQuery>
        </div>

        {/* Card controls (only shown if current user is viewing component) */}
        <div
          className={`${isCurrentUser ? " col-xs-3 text-center" : "hidden"}`}
        >
          <div className="pinned-recording-controls">
            <FontAwesomeIcon
              icon={faEllipsisV as IconProp}
              title="Delete"
              className="dropdown-toggle"
              id="pinnedRecordingControlsDropdown"
              data-toggle="dropdown"
              aria-haspopup="true"
              aria-expanded="true"
            />
            <ul
              className="dropdown-menu dropdown-menu-right"
              aria-labelledby="pinnedRecordingControlsDropdown"
            >
              {currentlyPinned && (
                <ListenControl
                  title="Unpin"
                  action={() => this.unpinRecording()}
                />
              )}
              <ListenControl
                title="Delete Pin"
                action={() => this.deletePin(pinnedRecording.row_id)}
              />
            </ul>
          </div>
        </div>
      </Card>
    );
  }
}
