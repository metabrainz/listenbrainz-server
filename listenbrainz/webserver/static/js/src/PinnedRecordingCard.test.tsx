import * as React from "react";
import { mount, shallow } from "enzyme";

import PinnedRecordingCard, {
  PinnedRecordingCardProps,
} from "./PinnedRecordingCard";
import * as utils from "./utils";
import APIServiceClass from "./APIService";
import GlobalAppContext from "./GlobalAppContext";

// Font Awesome generates a random hash ID for each icon everytime.
// Mocking Math.random() fixes this
// https://github.com/FortAwesome/react-fontawesome/issues/194#issuecomment-627235075
jest.spyOn(global.Math, "random").mockImplementation(() => 0);

const user = {
  id: 1,
  name: "name",
  auth_token: "auth_token",
};

const globalProps = {
  APIService: new APIServiceClass(""),
  currentUser: user,
  spotifyAuth: {},
  youtubeAuth: {},
};

const pinnedRecording: PinnedRecording = {
  blurb_content: "I LOVE",
  created: 1111111111,
  pinned_until: 9999999999,
  row_id: 1,
  recording_mbid: "98255a8c-017a-4bc7-8dd6-1fa36124572b",
  track_metadata: {
    artist_name: "Rick Astley",
    track_name: "Never Gonna Give You Up",
  },
};

const expiredPinnedRecording: PinnedRecording = {
  ...pinnedRecording,
  pinned_until: 1111122222,
};

const props: PinnedRecordingCardProps = {
  userName: user.name,
  pinnedRecording,
  isCurrentUser: true,
  newAlert: () => {},
};

describe("PinnedRecordingCard", () => {
  it("renders correctly", () => {
    const wrapper = mount<PinnedRecordingCard>(
      <PinnedRecordingCard {...props} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("calls renderPinTitle, renderPinDate, & renderBlurbContent", () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props, newAlert: jest.fn() }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    instance.renderPinTitle = jest.fn();
    instance.renderPinDate = jest.fn();
    instance.renderBlurbContent = jest.fn();

    instance.render();
    expect(instance.renderPinTitle).toHaveBeenCalledTimes(2);
    expect(instance.renderPinDate).toHaveBeenCalledTimes(2);
    expect(instance.renderBlurbContent).toHaveBeenCalledTimes(2);
  });
});

describe("determineIfCurrentlyPinned", () => {
  it("returns true when pinned_until > now", () => {
    const wrapper = mount<PinnedRecordingCard>(
      <PinnedRecordingCard {...props} />
    );
    const instance = wrapper.instance();
    expect(instance.determineIfCurrentlyPinned()).toBe(true);
  });

  it("returns false when pinned_until < now", () => {
    const wrapper = mount<PinnedRecordingCard>(
      <PinnedRecordingCard
        {...{ ...props, pinnedRecording: expiredPinnedRecording }}
      />
    );
    const instance = wrapper.instance();
    expect(instance.determineIfCurrentlyPinned()).toBe(false);
  });
});

describe("handleError", () => {
  it("calls newAlert", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <PinnedRecordingCard {...{ ...props, newAlert: jest.fn() }} />
    );
    const instance = wrapper.instance();

    instance.handleError("error");

    expect(instance.props.newAlert).toHaveBeenCalledTimes(1);
    expect(instance.props.newAlert).toHaveBeenCalledWith(
      "danger",
      "Error",
      "error"
    );
  });
});

describe("unpinRecording", () => {
  it("calls API, updates currentlyPinned in state, and calls newAlert", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props, newAlert: jest.fn() }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "unpinRecording");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("currentlyPinned")).toBeTruthy();

    await instance.unpinRecording();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("auth_token");

    expect(wrapper.state("currentlyPinned")).toBeFalsy();

    expect(instance.props.newAlert).toHaveBeenCalledTimes(1);
    expect(instance.props.newAlert).toHaveBeenCalledWith(
      "success",
      "You unpinned a recording.",
      "Rick Astley - Never Gonna Give You Up"
    );
  });

  it("does nothing if isCurrentUser is false", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props, isCurrentUser: false }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "unpinRecording");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("currentlyPinned")).toBeTruthy();
    await instance.unpinRecording();

    expect(spy).toHaveBeenCalledTimes(0);
    expect(wrapper.state("currentlyPinned")).toBeTruthy();
  });

  it("does nothing if CurrentUser.authtoken is not set", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider
        value={{
          ...globalProps,
          currentUser: { auth_token: undefined, name: "test" },
        }}
      >
        <PinnedRecordingCard {...props} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "unpinRecording");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("currentlyPinned")).toBeTruthy();
    await instance.unpinRecording();

    expect(spy).toHaveBeenCalledTimes(0);
    expect(wrapper.state("currentlyPinned")).toBeTruthy();
  });

  it("doesn't update currentlyPinned in state if status code is not 200", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "unpinRecording");
    spy.mockImplementation(() => Promise.resolve(201));

    expect(wrapper.state("currentlyPinned")).toBeTruthy();
    await instance.unpinRecording();

    expect(spy).toHaveBeenCalled();
    expect(wrapper.state("currentlyPinned")).toBeTruthy();
  });

  it("calls handleError if error is returned", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();
    instance.handleError = jest.fn();

    const error = new Error("error");
    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "unpinRecording");
    spy.mockImplementation(() => {
      throw error;
    });

    await instance.unpinRecording();
    expect(instance.handleError).toHaveBeenCalledTimes(1);
    expect(instance.handleError).toHaveBeenCalledWith(
      error,
      "Error while unpinning recording"
    );
  });
});

describe("deletePin", () => {
  it("calls API and updates isDeleted and currentlyPinned in state", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "deletePin");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("isDeleted")).toBeFalsy();
    expect(wrapper.state("currentlyPinned")).toBeTruthy();

    await instance.deletePin(pinnedRecording.row_id);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("auth_token", pinnedRecording.row_id);

    expect(wrapper.state("isDeleted")).toBeTruthy();
    expect(wrapper.state("currentlyPinned")).toBeFalsy();
  });

  it("does nothing if isCurrentUser is false", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props, isCurrentUser: false }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "deletePin");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("isDeleted")).toBeFalsy();
    await instance.deletePin(pinnedRecording.row_id);

    expect(spy).toHaveBeenCalledTimes(0);
    expect(wrapper.state("isDeleted")).toBeFalsy();
  });

  it("does nothing if CurrentUser.authtoken is not set", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider
        value={{
          ...globalProps,
          currentUser: { auth_token: undefined, name: "test" },
        }}
      >
        <PinnedRecordingCard {...props} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "deletePin");
    spy.mockImplementation(() => Promise.resolve(200));

    expect(wrapper.state("isDeleted")).toBeFalsy();
    await instance.deletePin(pinnedRecording.row_id);

    expect(spy).toHaveBeenCalledTimes(0);
    expect(wrapper.state("isDeleted")).toBeFalsy();
  });

  it("doesn't update currentlyPinned in state if status code is not 200", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();

    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "deletePin");
    spy.mockImplementation(() => Promise.resolve(201));

    expect(wrapper.state("isDeleted")).toBeFalsy();
    await instance.deletePin(pinnedRecording.row_id);

    expect(spy).toHaveBeenCalled();
    expect(wrapper.state("isDeleted")).toBeFalsy();
  });

  it("calls handleError if error is returned", async () => {
    const wrapper = mount<PinnedRecordingCard>(
      <GlobalAppContext.Provider value={globalProps}>
        <PinnedRecordingCard {...{ ...props }} />
      </GlobalAppContext.Provider>
    );
    const instance = wrapper.instance();
    instance.handleError = jest.fn();

    const error = new Error("error");
    const { APIService } = instance.context;
    const spy = jest.spyOn(APIService, "deletePin");
    spy.mockImplementation(() => {
      throw error;
    });

    await instance.deletePin(pinnedRecording.row_id);
    expect(instance.handleError).toHaveBeenCalledTimes(1);
    expect(instance.handleError).toHaveBeenCalledWith(
      error,
      "Error while deleting pin"
    );
  });
});
