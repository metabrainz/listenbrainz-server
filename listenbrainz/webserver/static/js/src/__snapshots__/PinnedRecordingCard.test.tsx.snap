// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinnedRecordingCard renders correctly 1`] = `
<PinnedRecordingCard
  isCurrentUser={true}
  newAlert={[Function]}
  pinnedRecording={
    Object {
      "blurb_content": "I LOVE",
      "created": 1111111111,
      "pinned_until": 9999999999,
      "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
      "row_id": 1,
      "track_metadata": Object {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      },
    }
  }
  userName="name"
>
  <ForwardRef
    className="pinned-recording-card row currently-pinned  "
  >
    <div
      className="pinned-recording-card row currently-pinned  "
      style={
        Object {
          "background": "#FFFFFF",
          "border": "1px solid #EEEEEE",
          "borderRadius": "8px",
          "boxShadow": "0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20)",
          "boxSizing": "border-box",
          "height": "100%",
        }
      }
    >
      <div
        className=" col-xs-9"
      >
        <MediaQuery
          minWidth={768}
        />
        <MediaQuery
          maxWidth={767}
        />
      </div>
      <div
        className=" col-xs-3 text-center"
      >
        <div
          className="pinned-recording-controls"
        >
          <FontAwesomeIcon
            aria-expanded="true"
            aria-haspopup="true"
            border={false}
            className="dropdown-toggle"
            data-toggle="dropdown"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  192,
                  512,
                  Array [],
                  "f142",
                  "M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z",
                ],
                "iconName": "ellipsis-v",
                "prefix": "fas",
              }
            }
            id="pinnedRecordingControlsDropdown"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title="Delete"
            transform={null}
          >
            <svg
              aria-expanded="true"
              aria-haspopup="true"
              aria-labelledby="svg-inline--fa-title-000000000000"
              className="svg-inline--fa fa-ellipsis-v fa-w-6 dropdown-toggle"
              data-icon="ellipsis-v"
              data-prefix="fas"
              data-toggle="dropdown"
              id="pinnedRecordingControlsDropdown"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-000000000000"
                style={Object {}}
              >
                Delete
              </title>
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </FontAwesomeIcon>
          <ul
            aria-labelledby="pinnedRecordingControlsDropdown"
            className="dropdown-menu dropdown-menu-right"
          >
            <ListenControl
              action={[Function]}
              title="Unpin"
            >
              <button
                onClick={[Function]}
                title="Unpin"
                type="button"
              >
                Unpin
              </button>
            </ListenControl>
            <ListenControl
              action={[Function]}
              title="Delete Pin"
            >
              <button
                onClick={[Function]}
                title="Delete Pin"
                type="button"
              >
                Delete Pin
              </button>
            </ListenControl>
          </ul>
        </div>
      </div>
    </div>
  </ForwardRef>
</PinnedRecordingCard>
`;
